core/src/test/java/com/linecorp/armeria/client/WebClientAdditionalAuthorityTest.java

```diff
@@ -0,0 +1,278 @@
+    @ParameterizedTest
+    void shouldRespectAuthorityInRequestHeaders(String protocol, String headerName) {
+        final BlockingWebClient client = WebClient.builder()
+                                                  .factory(clientFactory)
+                                                  .setHeader(headerName, "baz.com")
+                                                  .build()
+                                                  .blocking();
+
+        try (ClientRequestContextCaptor captor = Clients.newContextCaptor()) {
+            final RequestHeaders headers =
+                    RequestHeaders.builder(HttpMethod.GET, protocol + "://foo.com:" + fooServer.httpPort())
+                                  .authority("bar.com")
+                                  .build();
+
+            assertThat(client.execute(HttpRequest.of(headers)).contentUtf8())
+                    .isEqualTo("foo.com/bar.com");
+
+            assertThat(captor.get().endpoint().authority())
+                    .isEqualTo("foo.com:" + fooServer.httpPort());
+            assertThat(captor.get().log().whenComplete().join().requestHeaders().authority())
+                    .isEqualTo("bar.com");
+        }
+    }
+
+    @CsvSource({ "H1C, :authority", "H2C, :authority", "HTTP, :authority",
+                 "H1C, Host", "H2C, Host", "HTTP, Host" })
+    @ParameterizedTest
+    void shouldOverrideBaseAuthorityWithRequestAuthority(SessionProtocol protocol, String headerName) {
+        final BlockingWebClient client =
+                WebClient.builder()
+                         .factory(clientFactory)
+                         .addHeader(headerName, "foo.com:" + fooServer.httpPort())
+                         .build()
+                         .blocking();
+        try (ClientRequestContextCaptor captor = Clients.newContextCaptor()) {
+            final RequestHeaders headers =
+                    RequestHeaders.builder(HttpMethod.GET, "/")
+                                  .scheme(protocol)
+                                  .authority("bar.com:" + barServer.httpPort())
+                                  .build();
+
+            assertThat(client.execute(HttpRequest.of(headers)).contentUtf8())
+                    .isEqualTo("bar.com/bar.com:" + barServer.httpPort());
+            assertThat(captor.get().endpoint().authority())
+                    .isEqualTo("bar.com:" + barServer.httpPort());
+            assertThat(captor.get().log().whenComplete().join().requestHeaders().authority())
+                    .isEqualTo("bar.com:" + barServer.httpPort());
+        }
+    }
+
+    @CsvSource({ "H1C, :authority", "H2C, :authority", "HTTP, :authority",
+                 "H1C, Host", "H2C, Host", "HTTP, Host" })
+    @ParameterizedTest
+    void shouldOverrideRequestAuthorityWithAdditionalAuthority(String protocol, String headerName) {
+        try (SafeCloseable ignored = Clients.withHeader(headerName, "baz.com");
```

Comment:
Let's update the Javadoc of `Clients.withHeader(...)`. People might not notice that this uses additional headers internally so takes precedence over the headers from `RequestHeaders`.